File: python/design_patterns/decorator_pattern.py
from abc import ABC, abstractmethod

# Component Interface
class Beverage(ABC):
    @abstractmethod
    def get_description(self):
        pass

    @abstractmethod
    def cost(self):
        pass


# Concrete Components
class Espresso(Beverage):
    def get_description(self):
        return "Espresso"

    def cost(self):
        return 1.99

class HouseBlend(Beverage):
    def get_description(self):
        return "House Blend Coffee"

    def cost(self):
        return 0.89


# Decorator Base Class
class CondimentDecorator(Beverage, ABC):
    def __init__(self, beverage):
        self.beverage = beverage

    @abstractmethod
    def get_description(self):
        pass


# Concrete Decorators
class Mocha(CondimentDecorator):
    def get_description(self):
        return f"{self.beverage.get_description()}, Mocha"

    def cost(self):
        return self.beverage.cost() + 0.20

class Soy(CondimentDecorator):
    def get_description(self):
        return f"{self.beverage.get_description()}, Soy"

    def cost(self):
        return self.beverage.cost() + 0.15

class Whip(CondimentDecorator):
    def get_description(self):
        return f"{self.beverage.get_description()}, Whip"

    def cost(self):
        return self.beverage.cost() + 0.10


if __name__ == "__main__":
    # Order an Espresso with Mocha and Whip
    beverage = Espresso()
    print(f"{beverage.get_description()} ${beverage.cost():.2f}")

    beverage = Mocha(beverage)
    beverage = Whip(beverage)
    print(f"{beverage.get_description()} ${beverage.cost():.2f}")

    # Order a House Blend with Soy, Mocha, and Whip
    beverage2 = HouseBlend()
    beverage2 = Soy(beverage2)
    beverage2 = Mocha(beverage2)
    beverage2 = Whip(beverage2)
    print(f"{beverage2.get_description()} ${beverage2.cost():.2f}")

File: c++/design_patterns/decorator_pattern.cpp
#include <iostream>
#include <string>

// Abstract Component
class Beverage {
public:
    virtual ~Beverage() {}
    virtual std::string getDescription() const = 0;
    virtual double cost() const = 0;
};

// Concrete Components
class Espresso : public Beverage {
public:
    std::string getDescription() const {
        return "Espresso";
    }

    double cost() const {
        return 1.99;
    }
};

// Abstract Decorator
class CondimentDecorator : public Beverage {
public:
    virtual ~CondimentDecorator() {}
    virtual std::string getDescription() const = 0;  // Pure virtual, to be implemented by concrete decorators
};

// Concrete Decorators
class Mocha : public CondimentDecorator {
private:
    Beverage* beverage;  // Pointer to the wrapped Beverage

public:
    explicit Mocha(Beverage* b) : beverage(b) {}

    ~Mocha() {
        delete beverage;  // Clean up the wrapped object
    }

    std::string getDescription() const {
        return beverage->getDescription() + ", Mocha";
    }

    double cost() const {
        return 0.20 + beverage->cost();
    }
};

class Soy : public CondimentDecorator {
private:
    Beverage* beverage;

public:
    explicit Soy(Beverage* b) : beverage(b) {}

    ~Soy() {
        delete beverage;
    }

    std::string getDescription() const {
        return beverage->getDescription() + ", Soy";
    }

    double cost() const {
        return 0.15 + beverage->cost();
    }
};

class Whip : public CondimentDecorator {
private:
    Beverage* beverage;

public:
    explicit Whip(Beverage* b) : beverage(b) {}

    ~Whip() {
        delete beverage;
    }

    std::string getDescription() const {
        return beverage->getDescription() + ", Whip";
    }

    double cost() const {
        return 0.10 + beverage->cost();
    }
};

int main() {
    std::cout << "--------------------------" << std::endl;
    std::cout << "Decorator pattern" << std::endl;
    std::cout << "==========================" << std::endl;

    Beverage* beverage = new Espresso();
    std::cout << beverage->getDescription() << " $" << beverage->cost() << std::endl;

    beverage = new Mocha(beverage);
    std::cout << beverage->getDescription() << " $" << beverage->cost() << std::endl;

    beverage = new Whip(beverage);
    std::cout << beverage->getDescription() << " $" << beverage->cost() << std::endl;

    delete beverage;

    std::cout << "--------------------------" << std::endl << std::endl;
    return 0;
}
